Description: Enable QEMU-based cross build of Chromium
Author: Daniel Richard G. <skunk@iSKUNK.ORG>

The modifications in this patch allow our cross build of Chromium to run
foreign-arch build tooling via QEMU, in the manner of DEB_BUILD_OPTIONS =
crossbuildcanrunhostbinaries, without requiring binfmt-support to be set up
in the build environment.

Note that this is not the standard cross-build approach supported by the GN
build system. Instead of having separate configurations for host_cpu vs.
target_cpu, we use a single one (host_cpu == target_cpu == the foreign arch
we are targeting) and rely on QEMU for any foreign binary invocations
needed by the build. This turns out to be much easier to set up, since we
don't have to worry about two large parallel stacks of build dependencies.
(To say nothing of the fact that in a standard GN cross build, we end up
building a significant chunk of the codebase twice.)

Main objectives of this patch:

* Wherever a foreign-arch compiled tool may be executed, look at the
  HOST_EXEC_WRAPPER environment variable. If it is set, then split the
  value on whitespace, and prepend the resulting words to the tool
  invocation. This will usually be the appropriate QEMU user mode emulation
  binary, e.g. "qemu-armhf";

* Rust macros need to be handled a little differently, because they are
  dynamically loaded by rustc, and we are using the native-arch rustc.
  We thus special-case the build logic so that they are compiled for the
  native arch, unlike everything else.

Note that Debian and Chromium/GN use different nomenclature for the native
versus target architectures, which can make "host arch" ambiguous:

            | native arch    | foreign arch
  ----------+----------------+---------------
     Debian | DEB_BUILD_ARCH | DEB_HOST_ARCH
   Chromium | host_cpu       | target_cpu

(To make things even more confusing, there is also a DEB_TARGET_ARCH, but
it is relevant only to compilers and other tools that generate binaries.
There is no counterpart to it in the Chromium build system.)

See dpkg-architecture(1) for more information on the Debian side, and the
generate-ninja reference for details on host_* versus target_* variables:
https://gn.googlesource.com/gn/+/master/docs/reference.md#predefined_variables

Quick-and-dirty guide to performing a cross build:

(This is written to target armhf, but any other architecture supported by
the package may be substituted. Note that this should be run in a temporary
environment, as the first three steps will result in major changes to the
system. Also, while cross-building via sbuild(1) or the like should work as
a matter of course, the steps below represent a simpler, "manual" approach.)

1. # dpkg --add-architecture armhf

2. # apt-get update

3. # mk-build-deps -ir --host-arch armhf -Pcross /path/to/chromium-src/debian/control

4. (in source tree) $ dpkg-buildpackage -b --host-arch armhf -Pcross


--- a/build/config/compiler/BUILD.gn
+++ b/build/config/compiler/BUILD.gn
@@ -1702,7 +1702,12 @@ config("runtime_library") {
   # can be replaced by the C++ symbols. This config ensures the C++ symbols
   # exist and are strong in order to cause that replacement to occur by
   # explicitly linking in clang's compiler-rt library.
-  if (is_clang && !(is_a_target_toolchain && is_cronet_build)) {
+  if (is_clang && !(is_a_target_toolchain && is_cronet_build) &&
+      # Debian cross build: Only allow this for the target (foreign) arch.
+      # The clang builtins library is in libclang-rt-NN-dev, which (at this
+      # time) is not co-installable, so we don't have it available for both
+      # architectures.
+      current_cpu == host_cpu) {
     configs += [ "//build/config/clang:compiler_builtins" ]
   }
 
--- a/build/config/rust.gni
+++ b/build/config/rust.gni
@@ -408,6 +408,13 @@ if (toolchain_for_rust_host_build_tools)
   rust_macro_toolchain = "${host_toolchain}_for_rust_host_build_tools"
 }
 
+if (deb_build_cpu != "" && deb_build_cpu != host_cpu) {
+  # Debian cross build: Use the host (native) toolchain for Rust macro code,
+  # as it needs to be dlopen()ed by the native rustc.
+  rust_macro_toolchain =
+      "//build/toolchain/linux/unbundle:host_for_rust_host_build_tools"
+}
+
 # When this is true, a prebuilt Rust stdlib will be used. This has implications
 # such as that the panic strategy (unwind, abort) must match how the stdlib is
 # compiled, which is typically as unwind.
--- a/build/gn_run_binary.py
+++ b/build/gn_run_binary.py
@@ -22,6 +22,11 @@ if not os.path.isabs(path):
 # The rest of the arguments are passed directly to the executable.
 args = [path] + sys.argv[2:]
 
+# Debian cross-build support
+wrapper_env = os.environ.get('HOST_EXEC_WRAPPER')
+if wrapper_env:
+    args[0:0] = wrapper_env.split()
+
 ret = subprocess.call(args)
 if ret != 0:
     if ret <= -100:
--- a/build/toolchain/linux/unbundle/BUILD.gn
+++ b/build/toolchain/linux/unbundle/BUILD.gn
@@ -23,19 +23,24 @@ gcc_toolchain("default") {
 }
 
 gcc_toolchain("host") {
-  cc = getenv("BUILD_CC")
-  cxx = getenv("BUILD_CXX")
-  ar = getenv("BUILD_AR")
-  nm = getenv("BUILD_NM")
+  cc = getenv("CC")
+  cxx = getenv("CXX")
+  ar = getenv("AR")
+  nm = getenv("NM")
   ld = cxx
 
-  extra_cflags = getenv("BUILD_CFLAGS")
-  extra_cppflags = getenv("BUILD_CPPFLAGS")
-  extra_cxxflags = getenv("BUILD_CXXFLAGS")
-  extra_ldflags = getenv("BUILD_LDFLAGS")
+  extra_cflags = getenv("CFLAGS_FOR_BUILD")
+  extra_cppflags = getenv("CPPFLAGS_FOR_BUILD")
+  extra_cxxflags = getenv("CXXFLAGS_FOR_BUILD")
+  extra_ldflags = getenv("LDFLAGS_FOR_BUILD")
 
   toolchain_args = {
     current_cpu = current_cpu
+
+    if (deb_build_cpu != "") {
+      current_cpu = deb_build_cpu
+    }
+
     current_os = current_os
   }
 }
--- a/build/toolchain/toolchain.gni
+++ b/build/toolchain/toolchain.gni
@@ -44,6 +44,13 @@ declare_args() {
   clang_version = "24"
 }
 
+declare_args() {
+  # The GN CPU identifier corresponding to Debian's DEB_BUILD_ARCH.
+  # (In a Debian cross build, host_cpu/target_cpu are both set to the
+  # foreign architecture, so this specifies the native one.)
+  deb_build_cpu = ""
+}
+
 # Extension for shared library files (including leading dot).
 if (is_apple) {
   shlib_extension = ".dylib"
--- a/third_party/perfetto/gn/run_binary.py
+++ b/third_party/perfetto/gn/run_binary.py
@@ -18,5 +18,12 @@ This script exists because GN action onl
 import sys
 import subprocess
 
+# Debian cross-build support
+import os
+wrapper = []
+wrapper_env = os.environ.get('HOST_EXEC_WRAPPER')
+if wrapper_env:
+  wrapper = wrapper_env.split()
+
 if __name__ == '__main__':
-  sys.exit(subprocess.call(sys.argv[1:]))
+  sys.exit(subprocess.call(wrapper + sys.argv[1:]))
--- a/third_party/protobuf/src/google/protobuf/compiler/subprocess.cc
+++ b/third_party/protobuf/src/google/protobuf/compiler/subprocess.cc
@@ -309,7 +309,19 @@ void Subprocess::Start(const std::string
   ABSL_CHECK(pipe(stdin_pipe) != -1);
   ABSL_CHECK(pipe(stdout_pipe) != -1);
 
-  char* argv[2] = {portable_strdup(program.c_str()), nullptr};
+  char* argv[3];
+  int argc = 0;
+
+  // Debian cross-build support
+  char* wrapper_env = getenv("HOST_EXEC_WRAPPER");
+  if (wrapper_env) {
+    // TODO: split on whitespace
+    argv[argc++] = portable_strdup(wrapper_env);
+    search_mode = SEARCH_PATH;
+  }
+
+  argv[argc++] = portable_strdup(program.c_str());
+  argv[argc++] = nullptr;
 
   child_pid_ = fork();
   if (child_pid_ == -1) {
@@ -349,6 +361,8 @@ void Subprocess::Start(const std::string
     _exit(1);
   } else {
     free(argv[0]);
+    if (argc == 3)
+      free(argv[1]);
 
     close(stdin_pipe[0]);
     close(stdout_pipe[1]);
--- a/third_party/wayland/wayland_scanner_wrapper.py
+++ b/third_party/wayland/wayland_scanner_wrapper.py
@@ -14,8 +14,14 @@ import os.path
 import subprocess
 import sys
 
+# Debian cross-build support
+wrapper = []
+wrapper_env = os.environ.get('HOST_EXEC_WRAPPER')
+if wrapper_env:
+  wrapper = wrapper_env.split()
+
 def generate_code(wayland_scanner_cmd, scanner_args, path_in, path_out):
-  ret = subprocess.call([wayland_scanner_cmd, *scanner_args, path_in, path_out])
+  ret = subprocess.call(wrapper + [wayland_scanner_cmd, *scanner_args, path_in, path_out])
   if ret != 0:
     raise RuntimeError("wayland-scanner returned an error: %d" % ret)
 
@@ -38,7 +44,7 @@ def main(argv):
   protocols = options.protocols
   generator_type = options.generator_type
 
-  version = subprocess.check_output([cmd, "--version"],
+  version = subprocess.check_output(wrapper + [cmd, "--version"],
                                     stderr=subprocess.STDOUT).decode("utf-8")
   # The version is of the form "wayland-scanner 1.18.0\n"
   version = tuple([int(x) for x in version.strip().split(" ")[-1].split(".")])
--- a/tools/grit/grit/node/brotli_util.py
+++ b/tools/grit/grit/node/brotli_util.py
@@ -15,6 +15,11 @@ def SetBrotliCommand(brotli):
   global __brotli_executable
   __brotli_executable = brotli
 
+  # Debian cross-build support
+  import os
+  wrapper_env = os.environ.get('HOST_EXEC_WRAPPER')
+  if wrapper_env and __brotli_executable:
+    __brotli_executable[0:0] = wrapper_env.split()
 
 def BrotliCompress(data):
   if not __brotli_executable:
--- a/tools/protoc_wrapper/protoc_convert.py
+++ b/tools/protoc_wrapper/protoc_convert.py
@@ -10,6 +10,12 @@ of processing is not supported by GN.
 import argparse
 import subprocess
 
+# Debian cross-build support
+import os
+wrapper = []
+wrapper_env = os.environ.get('HOST_EXEC_WRAPPER')
+if wrapper_env:
+  wrapper = wrapper_env.split()
 
 def Main():
   parser = argparse.ArgumentParser()
@@ -30,7 +36,7 @@ def Main():
   stdout = open(args.outfile, 'w')
 
   subprocess.check_call(
-    [args.protoc] + passthrough_args, stdin=stdin, stdout=stdout
+    wrapper + [args.protoc] + passthrough_args, stdin=stdin, stdout=stdout
   )
 
 
--- a/tools/protoc_wrapper/protoc_wrapper.py
+++ b/tools/protoc_wrapper/protoc_wrapper.py
@@ -19,7 +19,7 @@ Features:
 
 from __future__ import print_function
 import argparse
-import os.path
+import os
 import subprocess
 import sys
 import tempfile
@@ -258,6 +258,16 @@ def main(argv):
     if not options.exclude_imports:
       protoc_cmd += ["--include_imports"]
 
+  # Debian cross-build support
+  wrapper_env = os.environ.get('HOST_EXEC_WRAPPER')
+  if wrapper_env:
+    protoc_cmd[0:0] = wrapper_env.split()
+    for arg in options.plugin, options.protoc_gen_js, options.protoc_gen_ts:
+      if arg and arg.endswith('.py'):
+        # Don't invoke a Python script via the wrapper
+        del os.environ['HOST_EXEC_WRAPPER']
+        break
+
   dependency_file_data = None
   if options.descriptor_set_out and options.descriptor_set_dependency_file:
     protoc_cmd += ['--dependency_out', options.descriptor_set_dependency_file]
--- a/ui/webui/resources/tools/generate_code_cache.py
+++ b/ui/webui/resources/tools/generate_code_cache.py
@@ -37,6 +37,12 @@ def main(argv):
   )
   out_file_suffix = '.code_cache'
 
+  # Debian cross-build support
+  wrapper = []
+  wrapper_env = os.environ.get('HOST_EXEC_WRAPPER')
+  if wrapper_env:
+    wrapper = wrapper_env.split()
+
   subprocess.run(
     [
       util_path,
--- a/v8/tools/run.py
+++ b/v8/tools/run.py
@@ -21,6 +21,12 @@ if cmd and cmd[0] == '--redirect-stdout'
   kwargs = dict(stdout=subprocess.PIPE)
   cmd = cmd[2:]
 
+# Debian cross-build support
+import os
+wrapper_env = os.environ.get('HOST_EXEC_WRAPPER')
+if wrapper_env:
+  cmd[0:0] = wrapper_env.split()
+
 process = subprocess.Popen(cmd, **kwargs)
 stdout, _ = process.communicate()
 if stdout_file:
